lib/src/main/java/com/diffplug/spotless/yaml/YamlJacksonStep.java

```diff
@@ -0,0 +1,146 @@
+
+			this.jarState = JarState.from(MAVEN_COORDINATE + jacksonVersion, provisioner);
+		}
+
+		FormatterFunc toFormatter() {
+			Class<?> jsonFactoryClass;
+			Class<?> yamlFactoryClass;
+			Class<?> objectMapperClass;
+
+			Class<?> serializationFeatureClass;
+			Method enableFeature;
+			Method disableFeature;
+
+			Method stringToObject;
+			Method objectToString;
+			try {
+				ClassLoader classLoader = jarState.getClassLoader();
+				jsonFactoryClass = classLoader.loadClass("com.fasterxml.jackson.core.JsonFactory");
+				yamlFactoryClass = classLoader.loadClass("com.fasterxml.jackson.dataformat.yaml.YAMLFactory");
+
+				objectMapperClass = classLoader.loadClass("com.fasterxml.jackson.databind.ObjectMapper");
+
+				// Configure the ObjectMapper
+				// https://github.com/FasterXML/jackson-databind#commonly-used-features
+				{
+					serializationFeatureClass = classLoader.loadClass("com.fasterxml.jackson.databind.SerializationFeature");
+					enableFeature = objectMapperClass.getMethod("enable", serializationFeatureClass);
+					disableFeature = objectMapperClass.getMethod("disable", serializationFeatureClass);
+				}
+
+				stringToObject = objectMapperClass.getMethod("readValue", String.class, Class.class);
+				objectToString = objectMapperClass.getMethod("writeValueAsString", Object.class);
+			} catch (ClassNotFoundException | NoSuchMethodException e) {
+				throw new IllegalStateException("There was a problem preparing org.json dependencies", e);
+			}
+
+			return s -> {
+				if (s.isEmpty()) {
+					return s;
+				}
+
+				Object yamlFactory = yamlFactoryClass.getConstructor().newInstance();
+				Object objectMapper = objectMapperClass.getConstructor(jsonFactoryClass).newInstance(yamlFactory);
+
+				for (String feature : enabledFeatures) {
+					// https://stackoverflow.com/questions/3735927/java-instantiating-an-enum-using-reflection
+					Object indentOutput = Enum.valueOf(serializationFeatureClass.asSubclass(Enum.class), feature);
+
+					enableFeature.invoke(objectMapper, indentOutput);
+				}
+
+				for (String feature : disabledFeatures) {
+					// https://stackoverflow.com/questions/3735927/java-instantiating-an-enum-using-reflection
+					Object indentOutput = Enum.valueOf(serializationFeatureClass.asSubclass(Enum.class), feature);
+
+					disableFeature.invoke(objectMapper, indentOutput);
+				}
+
+				return format(objectMapper, stringToObject, objectToString, s);
+			};
+		}
+
+		private String format(Object objectMapper, Method stringToObject, Method objectToString, String s)
+				throws IllegalAccessException, IllegalArgumentException {
+			try {
+				Object parsed = stringToObject.invoke(s, Object.class);
+				return (String) objectToString.invoke(objectMapper, parsed);
+			} catch (InvocationTargetException ex) {
+				throw new AssertionError("Unable to format YAML", ex.getCause());
+			}
+		}
+	}
+
+	private YamlJacksonStep() {
+		// cannot be directly instantiated
+	}
+}
```

Comment:
This reflection is fine, but justed to make sure you knew that compileOnly sourcesets are an option too. Lots of examples in 

- https://github.com/diffplug/spotless/issues/524

Reply:
I read the theory (and one of the example PR) but I was not able to grasp the various implications and implementation needs.

...

After more read, https://github.com/diffplug/spotless/commit/0d83dca8cc087a2773d1202ccf84e58a54d65631#diff-c8bd815c0e50351750df4c8fa7881aaea385ef2fe0ee033b1625361164570f4d seemed a good reference for my understanding.

I may switch to this format later (this Step, or more Steps to come).

FYI, one of my main motivation here is integrating https://github.com/solven-eu/cleanthat/blob/master/java/src/main/java/eu/solven/cleanthat/java/mutators/RulesJavaMutator.java as a java step in Spotless, and later leverage Spotless in https://github.com/solven-eu/cleanthat/tree/IntroduceSpotLess

Reply:
> I may switch to this format later (this Step, or more Steps to come).

Sounds good, no preference from me.

>  my main motivation here is integrating ... cleanthat

Cool!

FYI, I'm planning a release this afternoon. This PR looks like it's pretty close to making that train, but it won't be long before the next one comes so non rush :)

Reply:
I just spotted the lack of `Json json` in `AbstractSpotlessMojo`. Hence, I'm not confident in JSON maven integration (and it is unclear to me why it was not captured by the unitTests).

Reply:
Ha! I don't know how those tests are passing either! They're in the CI logs

> com.diffplug.spotless.maven.json.JsonTest
>
>  Test testFormatJson_WithGson_defaultConfig() PASSED (1s)
>  Test testFormatJson_WithSimple_defaultConfig() PASSED (1.1s)

Feel free to fix that in this PR if you'd like. I guess I'll hold off on releasing until we get to the bottom of it. I don't have time to dig in to this today, but I'll take a look later in the week.

Reply:
I fixed 2 weird things related to JSON tests. They still pass.

But my Yaml tests do not pass (like if the process is not applied at all). I can not find why.

Reply:
Even with `writePom();` instead of `writePomWithJsonSteps("<gson/>");` the JSONs are formatted in the unitTests. Something funky is going-on...

Reply:
I fear I have picked no op test cases in json... No battery anymore, I ll pursue tomorrow.

Reply:
@nedtwigg I added not trivial json tests and it confirms JSON mvn plugin is not formatting anything. Sorry for that.

I suppose the issue with json and yaml (in mvn) is the same. Any pointer/hint about the potential issue would be beneficial.

Reply:
I think the problem is probably that the test is not setting a target. Add the default `**/*` target back in, does that fix it?

Reply:
Forcing the deployement of the plugin in maven

    ../gradlew publishToMavenLocal

and running the mojo locally

    mvn com.diffplug.spotless:spotless-maven-plugin:2.30.0-SNAPSHOT:check -X

I'm able to see some details.

With

    mvn help:describe -Dplugin=com.diffplug.spotless:spotless-maven-plugin:2.30.0-SNAPSHOT -Dgoal=check -Ddetail
    
We see it lacks the JSON and YAML sections.

---

I'm not sure to get the target issue.

I tried following the pom test:

	@Test
	public void testSortPomWithDefaultConfig() throws Exception {
		writePomWithPomSteps("<sortPom/>");

		setFile("pom_test.xml").toResource("pom/pom_dirty.xml");
		mavenRunner().withArguments("spotless:apply").runNoError().error();
		assertFile("pom_test.xml").sameAsResource("pom/pom_clean_default.xml");
	}
	
I also changed from 

	protected void writePomWithJsonSteps(String... steps) throws IOException {
		writePom(groupWithSteps("json", including("**/*.json"), steps));
	}
	
to 
	
		protected void writePomWithJsonSteps(String... steps) throws IOException {
		writePom(groupWithSteps("json", including("**/*"), steps));
	}
	
but it still does not work.

Reply:
@lutovich do you have any hints for us in this case of the missing language extension? The PR below seems like the simplest example of adding a format

- https://github.com/diffplug/spotless/pull/857

and this PR does all of those things...

Reply:
I was using `../gradlew test --tests '*yaml*'` to test Yaml related stuff. it appears to be skipping some issues, as the CI reports more with `./gradlew :plugin-maven:check --build-cache`. Maybe both YAML and JSON both have an issue in the reflection code.

Reply:
Ok. It looks like the issue was exclusively in my local setup/commands. Which is good news. I'll try to complete the YAML integration asap.